transports/webtransport/src/transport.rs

```diff
@@ -0,0 +1,682 @@
+    }
+
+    socket.bind(&socket_addr.into())?;
+
+    Ok(socket.into())
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub(crate) enum SocketFamily {
+    Ipv4,
+    Ipv6,
+}
+
+impl SocketFamily {
+    fn is_same(a: &IpAddr, b: &IpAddr) -> bool {
+        matches!(
+            (a, b),
+            (IpAddr::V4(_), IpAddr::V4(_)) | (IpAddr::V6(_), IpAddr::V6(_))
+        )
+    }
+}
+
+impl From<IpAddr> for SocketFamily {
+    fn from(ip: IpAddr) -> Self {
+        match ip {
+            IpAddr::V4(_) => SocketFamily::Ipv4,
+            IpAddr::V6(_) => SocketFamily::Ipv6,
+        }
+    }
+}
+
+/// Turn an [`IpAddr`] reported by the interface watcher into a
+/// listen-address for the endpoint.
+///
+/// For this, the `ip` is combined with the port that the endpoint
+/// is actually bound.
+///
+/// Returns `None` if the `ip` is not the same socket family as the
+/// address that the endpoint is bound to.
+fn ip_to_listen_addr(
+    endpoint_addr: &SocketAddr,
+    ip: IpAddr,
+    hashes: &[CertHash],
+) -> Option<Multiaddr> {
+    // True if either both addresses are Ipv4 or both Ipv6.
+    if !SocketFamily::is_same(&endpoint_addr.ip(), &ip) {
+        return None;
+    }
+    let socket_addr = SocketAddr::new(ip, endpoint_addr.port());
+    Some(socketaddr_to_multiaddr_with_hashes(&socket_addr, hashes))
+}
+
+/// Turns an IP address and port into the corresponding WebTransport multiaddr.
+fn socketaddr_to_multiaddr(socket_addr: &SocketAddr) -> Multiaddr {
+    Multiaddr::empty()
+        .with(socket_addr.ip().into())
+        .with(Protocol::Udp(socket_addr.port()))
+        .with(Protocol::QuicV1)
+        .with(Protocol::WebTransport)
+}
+
+fn socketaddr_to_multiaddr_with_hashes(socket_addr: &SocketAddr, hashes: &[CertHash]) -> Multiaddr {
+    let mut res = socketaddr_to_multiaddr(socket_addr);
+
+    if !hashes.is_empty() {
+        let mut vec = hashes.to_owned();
+        res = res.with(Protocol::Certhash(
+            vec.pop().expect("Gets the last element"),
+        ));
+        if !vec.is_empty() {
+            res = res.with(Protocol::Certhash(
+                vec.pop().expect("Gets the last element"),
+            ));
+        };
+    }
+
+    res
+}
+
+/// Tries to turn a Webtransport multiaddress into a UDP [`SocketAddr`]. Returns None if the format
+/// of the multiaddr is wrong.
+fn multiaddr_to_socketaddr(addr: &Multiaddr) -> Option<(SocketAddr, Option<PeerId>)> {
+    let mut iter = addr.iter();
+    let proto1 = iter.next()?;
+    let proto2 = iter.next()?;
+    let proto3 = iter.next()?;
+
+    if !matches!(proto3, Protocol::QuicV1) {
+        tracing::error!("Cannot listen on a non QUIC address {addr}");
+        return None;
+    }
+
+    let mut peer_id = None;
+    let mut is_webtransport = false;
+    for proto in iter {
+        match proto {
+            Protocol::P2p(id) => {
+                peer_id = Some(id);
+            }
+            Protocol::WebTransport => {
```

Comment:
```suggestion
            Protocol::WebTransport if !is_webtransport => {
```

Prevent duplicate `/webtransport` proto.